toolchain/check/testdata/impl/no_prelude/impl_cycle.carbon

```diff
@@ -194,3 +194,65 @@ fn F() {
   Compare({} as C, {} as D);
   Compare({} as D, {} as C);
 }
+
+// --- deduce_cycle_without_symbolic.carbon
+library "[[@TEST_NAME]]";
+
+interface First {}
+interface Second {}
+
+// This impl is looked at for "does T impl Second" first, which requires a
+// recursive query in deduction that is the same "does T impl Second", which can
+// land us back looking at this impl again. So we require a way to eliminate
+// this impl without doing deduction in order to avoid the cycle.
```

Comment:
This feels like it ends on a cliffhanger -- evidently we found such a way, since there's no diagnostic, but what is it?

Reply:
Yes, this is the early out here: https://github.com/carbon-language/carbon-lang/blob/186ca0e5058ad033573aead30ddbb6963e7eafb6/toolchain/check/impl_lookup.cpp#L119-L126

Fixing the TODO below it will make the next test pass eventually.